// @dhis2/pwa/build/cjs/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "OfflineInterface", {
  enumerable: true,
  get: function () {
    return _offlineInterface.OfflineInterface;
  }
});
Object.defineProperty(exports, "REGISTRATION_STATE_ACTIVE", {
  enumerable: true,
  get: function () {
    return _registration.REGISTRATION_STATE_ACTIVE;
  }
});
Object.defineProperty(exports, "REGISTRATION_STATE_FIRST_ACTIVATION", {
  enumerable: true,
  get: function () {
    return _registration.REGISTRATION_STATE_FIRST_ACTIVATION;
  }
});
Object.defineProperty(exports, "REGISTRATION_STATE_UNREGISTERED", {
  enumerable: true,
  get: function () {
    return _registration.REGISTRATION_STATE_UNREGISTERED;
  }
});
Object.defineProperty(exports, "REGISTRATION_STATE_WAITING", {
  enumerable: true,
  get: function () {
    return _registration.REGISTRATION_STATE_WAITING;
  }
});
Object.defineProperty(exports, "checkForSWUpdateAndReload", {
  enumerable: true,
  get: function () {
    return _registration.checkForSWUpdateAndReload;
  }
});
Object.defineProperty(exports, "checkForUpdates", {
  enumerable: true,
  get: function () {
    return _registration.checkForUpdates;
  }
});
Object.defineProperty(exports, "getBaseUrlByAppName", {
  enumerable: true,
  get: function () {
    return _baseUrlDb.getBaseUrlByAppName;
  }
});
Object.defineProperty(exports, "getRegistrationState", {
  enumerable: true,
  get: function () {
    return _registration.getRegistrationState;
  }
});
Object.defineProperty(exports, "setBaseUrlByAppName", {
  enumerable: true,
  get: function () {
    return _baseUrlDb.setBaseUrlByAppName;
  }
});
Object.defineProperty(exports, "setUpServiceWorker", {
  enumerable: true,
  get: function () {
    return _setUpServiceWorker.setUpServiceWorker;
  }
});

var _setUpServiceWorker = require("./service-worker/set-up-service-worker.js");

var _offlineInterface = require("./offline-interface/offline-interface.js");

var _registration = require("./lib/registration.js");

var _baseUrlDb = require("./lib/base-url-db.js");
