// @dhis2/app-service-offline/build/cjs/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "CacheableSection", {
  enumerable: true,
  get: function () {
    return _cacheableSection.CacheableSection;
  }
});
Object.defineProperty(exports, "OfflineProvider", {
  enumerable: true,
  get: function () {
    return _offlineProvider.OfflineProvider;
  }
});
Object.defineProperty(exports, "clearSensitiveCaches", {
  enumerable: true,
  get: function () {
    return _clearSensitiveCaches.clearSensitiveCaches;
  }
});
Object.defineProperty(exports, "useCacheableSection", {
  enumerable: true,
  get: function () {
    return _cacheableSection.useCacheableSection;
  }
});
Object.defineProperty(exports, "useCachedSections", {
  enumerable: true,
  get: function () {
    return _cacheableSectionState.useCachedSections;
  }
});
Object.defineProperty(exports, "useDhis2ConnectionStatus", {
  enumerable: true,
  get: function () {
    return _dhis2ConnectionStatus.useDhis2ConnectionStatus;
  }
});
Object.defineProperty(exports, "useOnlineStatus", {
  enumerable: true,
  get: function () {
    return _networkStatus.useNetworkStatus;
  }
});
Object.defineProperty(exports, "useOnlineStatusMessage", {
  enumerable: true,
  get: function () {
    return _onlineStatusMessage.useOnlineStatusMessage;
  }
});

var _offlineProvider = require("./lib/offline-provider");

var _cacheableSection = require("./lib/cacheable-section");

var _cacheableSectionState = require("./lib/cacheable-section-state");

var _networkStatus = require("./lib/network-status");

var _onlineStatusMessage = require("./lib/online-status-message");

var _clearSensitiveCaches = require("./lib/clear-sensitive-caches");

var _dhis2ConnectionStatus = require("./lib/dhis2-connection-status");
