// @dhis2/app-adapter/build/cjs/components/ConfirmUpdateModal.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ConfirmUpdateModal = ConfirmUpdateModal;

var _ui = require("@dhis2/ui");

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _locales = _interopRequireDefault(require("../locales"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function ConfirmUpdateModal(_ref) {
  let {
    clientsCount,
    onCancel,
    onConfirm
  } = _ref;
  return /*#__PURE__*/_react.default.createElement(_ui.Modal, {
    position: "middle"
  }, /*#__PURE__*/_react.default.createElement(_ui.ModalTitle, null, _locales.default.t('Save your data')), /*#__PURE__*/_react.default.createElement(_ui.ModalContent, null, clientsCount ? _locales.default.t("Updating will reload all {{n}} open instances of this app, and any unsaved data will be lost. Save any data you need to, then click 'Reload' when ready.", {
    n: clientsCount
  }) : // Fallback if clientsCount is unavailable:
  _locales.default.t("Updating will reload all open instances of this app, and any unsaved data will be lost. Save any data you need to, then click 'Reload' when ready.")), /*#__PURE__*/_react.default.createElement(_ui.ModalActions, null, /*#__PURE__*/_react.default.createElement(_ui.ButtonStrip, {
    end: true
  }, /*#__PURE__*/_react.default.createElement(_ui.Button, {
    onClick: onCancel
  }, _locales.default.t('Cancel')), /*#__PURE__*/_react.default.createElement(_ui.Button, {
    destructive: true,
    onClick: onConfirm
  }, _locales.default.t('Reload')))));
}

ConfirmUpdateModal.propTypes = {
  clientsCount: _propTypes.default.number,
  onCancel: _propTypes.default.func,
  onConfirm: _propTypes.default.func
};
