// @dhis2/app-adapter/build/cjs/components/ConnectedHeaderBar.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ConnectedHeaderBar = ConnectedHeaderBar;

var _appRuntime = require("@dhis2/app-runtime");

var _ui = require("@dhis2/ui");

var _react = _interopRequireDefault(require("react"));

var _usePWAUpdateState = require("../utils/usePWAUpdateState");

var _ConfirmUpdateModal = require("./ConfirmUpdateModal");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * Check for SW updates or a first activation, displaying an update notification
 * message in the HeaderBar profile menu. When an update is applied, if there are
 * multiple tabs of this app open, there's anadditional warning step because all
 * clients of the service worker will reload when there's an update, which may
 * cause data loss.
 */
function ConnectedHeaderBar() {
  const {
    appName
  } = (0, _appRuntime.useConfig)();
  const {
    updateAvailable,
    confirmReload,
    confirmationRequired,
    clientsCount,
    onConfirmUpdate,
    onCancelUpdate
  } = (0, _usePWAUpdateState.usePWAUpdateState)();
  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(_ui.HeaderBar, {
    appName: appName,
    updateAvailable: updateAvailable,
    onApplyAvailableUpdate: confirmReload
  }), confirmationRequired ? /*#__PURE__*/_react.default.createElement(_ConfirmUpdateModal.ConfirmUpdateModal, {
    clientsCount: clientsCount,
    onConfirm: onConfirmUpdate,
    onCancel: onCancelUpdate
  }) : null);
}
