// @dhis2-ui/calendar/build/cjs/stories/calendar.stories.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.WithAnyCalendar = exports.Nepali = exports.Ethiopic = void 0;

var _react = _interopRequireDefault(require("react"));

var _calendar = require("../calendar/calendar.js");

var _calendarStoryWrapper = require("./calendar-story-wrapper.js");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

const subtitle = "[Experimental] Calendar component is useful for creating a variety of calendars including Ethiopic, Islamic etc..";
const description = "\nUse a Calendar where there is a need to let the user select a date.\n\nNote that it requires a parent, like [Box](../?path=/docs/layout-box--default), to define its size.\n\n```js\nimport { Calendar } from '@dhis2/ui'\n```\n";
var _default = {
  title: 'Calendar',
  component: _calendar.Calendar,
  parameters: {
    componentSubtitle: subtitle,
    docs: {
      description: {
        component: description
      }
    }
  }
};
exports.default = _default;

const Ethiopic = args => {
  return /*#__PURE__*/_react.default.createElement(_calendarStoryWrapper.CalendarStoryWrapper, _extends({
    calendar: "ethiopic",
    locale: "am-et",
    numberingSystem: "ethi",
    date: "2014-04-05" // 13th of October 2022
    ,
    timeZone: "Africa/Khartoum"
  }, args));
};

exports.Ethiopic = Ethiopic;

const Nepali = args => {
  return /*#__PURE__*/_react.default.createElement(_calendarStoryWrapper.CalendarStoryWrapper, _extends({
    calendar: "nepali",
    locale: "ne-NP",
    date: "2079-06-29" // 13th of October 2022
    ,
    timeZone: "Africa/Khartoum"
  }, args));
};

exports.Nepali = Nepali;

const WithAnyCalendar = args => {
  return /*#__PURE__*/_react.default.createElement(_calendarStoryWrapper.CalendarStoryWrapper, _extends({
    calendar: "gregory",
    locale: "en-GB"
  }, args));
};

exports.WithAnyCalendar = WithAnyCalendar;
