// @dhis2-ui/calendar/build/es/stories/calendar.stories.js
function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

import React from 'react';
import { Calendar } from '../calendar/calendar.js';
import { CalendarStoryWrapper } from './calendar-story-wrapper.js';
const subtitle = "[Experimental] Calendar component is useful for creating a variety of calendars including Ethiopic, Islamic etc..";
const description = "\nUse a Calendar where there is a need to let the user select a date.\n\nNote that it requires a parent, like [Box](../?path=/docs/layout-box--default), to define its size.\n\n```js\nimport { Calendar } from '@dhis2/ui'\n```\n";
export default {
  title: 'Calendar',
  component: Calendar,
  parameters: {
    componentSubtitle: subtitle,
    docs: {
      description: {
        component: description
      }
    }
  }
};
export const Ethiopic = args => {
  return /*#__PURE__*/React.createElement(CalendarStoryWrapper, _extends({
    calendar: "ethiopic",
    locale: "am-et",
    numberingSystem: "ethi",
    date: "2014-04-05" // 13th of October 2022
    ,
    timeZone: "Africa/Khartoum"
  }, args));
};
export const Nepali = args => {
  return /*#__PURE__*/React.createElement(CalendarStoryWrapper, _extends({
    calendar: "nepali",
    locale: "ne-NP",
    date: "2079-06-29" // 13th of October 2022
    ,
    timeZone: "Africa/Khartoum"
  }, args));
};
export const WithAnyCalendar = args => {
  return /*#__PURE__*/React.createElement(CalendarStoryWrapper, _extends({
    calendar: "gregory",
    locale: "en-GB"
  }, args));
};
