// @dhis2/app-runtime/build/cjs/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "CacheableSection", {
  enumerable: true,
  get: function () {
    return _appServiceOffline.CacheableSection;
  }
});
Object.defineProperty(exports, "CustomDataProvider", {
  enumerable: true,
  get: function () {
    return _appServiceData.CustomDataProvider;
  }
});
Object.defineProperty(exports, "DataMutation", {
  enumerable: true,
  get: function () {
    return _appServiceData.DataMutation;
  }
});
Object.defineProperty(exports, "DataProvider", {
  enumerable: true,
  get: function () {
    return _appServiceData.DataProvider;
  }
});
Object.defineProperty(exports, "DataQuery", {
  enumerable: true,
  get: function () {
    return _appServiceData.DataQuery;
  }
});
Object.defineProperty(exports, "FetchError", {
  enumerable: true,
  get: function () {
    return _appServiceData.FetchError;
  }
});
Object.defineProperty(exports, "Provider", {
  enumerable: true,
  get: function () {
    return _Provider.Provider;
  }
});
Object.defineProperty(exports, "clearSensitiveCaches", {
  enumerable: true,
  get: function () {
    return _appServiceOffline.clearSensitiveCaches;
  }
});
Object.defineProperty(exports, "useAlert", {
  enumerable: true,
  get: function () {
    return _appServiceAlerts.useAlert;
  }
});
Object.defineProperty(exports, "useAlerts", {
  enumerable: true,
  get: function () {
    return _appServiceAlerts.useAlerts;
  }
});
Object.defineProperty(exports, "useCacheableSection", {
  enumerable: true,
  get: function () {
    return _appServiceOffline.useCacheableSection;
  }
});
Object.defineProperty(exports, "useCachedSections", {
  enumerable: true,
  get: function () {
    return _appServiceOffline.useCachedSections;
  }
});
Object.defineProperty(exports, "useConfig", {
  enumerable: true,
  get: function () {
    return _appServiceConfig.useConfig;
  }
});
Object.defineProperty(exports, "useDataEngine", {
  enumerable: true,
  get: function () {
    return _appServiceData.useDataEngine;
  }
});
Object.defineProperty(exports, "useDataMutation", {
  enumerable: true,
  get: function () {
    return _appServiceData.useDataMutation;
  }
});
Object.defineProperty(exports, "useDataQuery", {
  enumerable: true,
  get: function () {
    return _appServiceData.useDataQuery;
  }
});
Object.defineProperty(exports, "useDhis2ConnectionStatus", {
  enumerable: true,
  get: function () {
    return _appServiceOffline.useDhis2ConnectionStatus;
  }
});
Object.defineProperty(exports, "useOnlineStatus", {
  enumerable: true,
  get: function () {
    return _appServiceOffline.useOnlineStatus;
  }
});
Object.defineProperty(exports, "useOnlineStatusMessage", {
  enumerable: true,
  get: function () {
    return _appServiceOffline.useOnlineStatusMessage;
  }
});
Object.defineProperty(exports, "useTimeZoneConversion", {
  enumerable: true,
  get: function () {
    return _appServiceConfig.useTimeZoneConversion;
  }
});

var _appServiceData = require("@dhis2/app-service-data");

var _appServiceConfig = require("@dhis2/app-service-config");

var _appServiceAlerts = require("@dhis2/app-service-alerts");

var _appServiceOffline = require("@dhis2/app-service-offline");

var _Provider = require("./Provider");
